### PR TITLE
MAGN-7390 Category list in Library is getting changed after placing Custom node from that category

### PR DESCRIPTION
**Pull request by aosyatnik**
#### Link to YouTrack
[MAGN-7390](http://adsk-oss.myjetbrains.com/youtrack/issue/MAGN-7390) [Regression:0.8.0] Category list in Library is getting changed after placing Custom node from that category
*Priority: Show-stopper*

##### Reason of bug
For the first time user clicks on custom node, it's created custom workspace. As result of it search model updates itself. That means delete node and add this node again.
Removing node is made so that:
if node is removed, SearchVM tries to load all nodes, that are not part of class, into one class.
This works pretty good, when category from which item is deleted is not root category.

So, as soon as node is deleted SearchVM looks though its' nodes, that are not part of class. SearchVM creates class and adds it at the top.
That's why it appears like that:
image
The solution is to check whether parent item is root tree.

##### Another bug
During trying many other possible variants of placing members, I found one more bug.
Let's say we have next structure:
image

We have loop, where for each element in top category, we check if element name is bigger than name of adding item. If name is bigger, insert item before this element.
Since element is member, then found index + number of categories.
image
image

BUT, as we know, letter B should  be after A. So right place for BMember is after AMember.
image

I've added fix and test for that case.

### Declarations

- [ ] The code base is in a better state after this PR
- [ ] The level of testing this PR includes is appropriate

### Reviewers

- [ ] @Benglin 

### FYIs

@riteshchandawar

### CODE REVIEW

**Comment by riteshchandawar** [comment]
@aosyatnik I tried your fix, nodes from Migration category seems to be fixed, but if you have a custom node installed from Packages and if it has only one Custom Node inside that category then it is still jumping to top level in the Library.

Install "GetMidPoint" and "LoftWithHole" nodes from Packages and play with it.

If you can't reproduce then let me know I will attach video with all details.

Thanks,
Ritesh

**Comment by aosyatnik** [comment]
Hi, @riteshchandawar . Thanks for testing it. I've tried and I can't reproduce exactly this behavior. Top category with just 1 custom node doesn't jump at the first place for me.
Yes, there is 1 little bug: it collapses(closes), that is because node is deleted and then added again.
But category doesn't jump to the top.

**Comment by Benglin** [comment]
Hi @aosyatnik, since there's a defect I'm going to look at this after a while. I'm wondering if there's a way to not perform the reordering if it is a root-level insertion? The code changes begin to look a little... complex :)

**Comment by Benglin** [comment]
Hi @aosyatnik, I had a chance to debugging through this for a while, and I noticed something we could have done better. Looking at `SearchLibrary.Update` method, it currently does a `Remove` followed by `Add`, which resulted in this defect in the first place. I'm wondering if the following can be done instead? The `Update` operation should just result in updating an existing element, or adding of that element if it does not exist:

```csharp
class SearchLibrary
{
    public void Update(TEntry entry)
    {
        Dictionary<string, double> keys;
        if (entryDictionary.TryGetValue(entry, out keys))
        {
            keys[entry.Name.ToLower()] = 1.0;
            keys[entry.Description.ToLower()] = 0.1;
            
            foreach (var tag in entry.SearchTags.Select(x => x.ToLower()))
                keys[tag] = 0.5;

            OnEntryUpdated(entry); // NOTE: You need to add this.
            return; // Entry updated.
        }
        
        // Entry could not be found, add it into the search collection.
        Add(entry);
    }
}
```

What are your thoughts on this one? Please let me know.

File: src/DynamoCore/Search/SearchDictionary.cs [hunk 1]
```diff
@@ -12,7 +12,7 @@ namespace Dynamo.Search
     /// </summary>
     public class SearchDictionary<V>
     {
-        private readonly Dictionary<V, Dictionary<string, double>> entryDictionary =
+        protected readonly Dictionary<V, Dictionary<string, double>> entryDictionary =
```
Comment (aosyatnik) [inline src/DynamoCore/Search/SearchDictionary.cs]: I have to change from `private` to `protected`, because I need to use it in `SearchLibrary`.

**Comment by aosyatnik** [comment]
Hi, @Benglin , thank you for great idea. I tried to implement it. I couldn't find any bugs caused by these changes. Seems everything works properly. Please, take a look.

**Comment by Benglin** [comment]
Thanks a lot @aosyatnik, now this LGTM! Let's get this in for further testing by @riteshchandawar :)

**Pull request merged**